EIPS/eip-6492.md

```diff
@@ -0,0 +1,143 @@
+
+
+Quoting [ERC-1271](./eip-1271.md),
+> `isValidSignature` can call arbitrary methods to validate a given signature, which could be context dependent (e.g. time based or state based), EOA dependent (e.g. signers authorization level within smart wallet), signature scheme Dependent (e.g. ECDSA, multisig, BLS), etc. 
+>
+> This function should be implemented by contracts which desire to sign messages (e.g. smart contract wallets, DAOs, multisignature wallets, etc.) Applications wanting to support contract signatures should call this method if the signer is a contract.
+
+
+We use the same `isValidSignature` function, but we add a new wrapper signature format, that signing contracts MAY use before they're deployed, in order to allow support for verification.
+
+The signature verifier MUST perform a contract deployment before attempting to call `isValidSignature` if the wrapper signature format is detected.
+
+The wrapper format is detected by checking if the signature ends in `magicBytes`, which MUST be defined as `0x69696969`.
+
+It is RECOMMENDED to use this ERC with CREATE2 contracts, as their deploy address is always predictable.
+
+### Signer side
+
+The signer will normally be a contract wallet, but it could be any other counterfactual contract as well.
+
+- If the contract is deployed, produce a normal [ERC-1271](./eip-1271.md) signature
+- If the contract is not deployed yet, wrap the signature as follows: `concat(abi.encodePacked((create2Factory, factoryCalldata, originalERC1271Signature), (address, bytes32, bytes, bytes, bytes)), magicBytes)`
+
+Note that we're passing `factoryCalldata` instead of `salt` and `bytecode`. We do this in order to make verification compliant with any factory interface. We do not need to calculate the address based on  `create2Factory`/`salt`/`bytecode`, because [ERC-1271](./eip-1271.md) verification presumes we already know the account address we're verifying the signature for.
+
+### Verifier side
+
+Full signature verification MUST be performed in the following order:
+
+- check if there's contract code at the address. If so perform [ERC-1271](./eip-1271.md) verification as usual by invoking `isValidSignature`
+- check if the signature ends with magic bytes, in which case do an `eth_call` to a multicall contract that will call the factory first with the `factoryCalldata` and deploy the contract; Then, call `contract.isValidSignature` as usual
+- If all this fails, try `ecrecover` verification
+
+## Rationale
+
+We believe that wrapping the signature in a way that allows to pass the deploy data is the only clean way to implement this, as it's completely contract agnostic, but also easy to verify.
+
+The wrapper format ends in `magicBytes`, which ends with a `0x69`, which makes it is impossible for it to collide with a valid `ecrecover` signature if packed in the `r,s,v` format, as `0x69` is not a valid value for `v`.
```

Comment:
In general, a prefix is used to mark data, not suffix.
also it is customary to use the EIP number as ID, So I would suggest using "0x64926492" as a magic value.

Reply:
the suffix is used to make it impossible for the detection to collide with a valid r,s,v signature, with the presumption that the de-facto standard is that v is packed last

As for the magic value, that makes sense